### PR TITLE
Add a method field to trails

### LINKED ISSUE #10: Add a method field to trails
- [ ] `metadata` field in `trails` map should be of type `metadata() :: #{method() => map()}.`
- [ ] `method() :: get | put | post ...`
- [ ] add inaka's default dialyzer options and run `make dialyzer`

### PR DESCRIPTION
**Pull request by davecaos**
Fix #10 issue.

### CODE REVIEW

**Comment by gadgetci** [comment]
According to **Elvis**:

> Dependency 'cowlib' is not using appropriate protocol, please change this to something like 'git://.*'

**Comment by gadgetci** [comment]
According to **Elvis**:

> Dependency 'ranch' is not using appropriate protocol, please change this to something like 'git://.*'

File: src/trails.erl [hunk 1]
```diff
@@ -13,35 +13,52 @@
...
+   , constraints => cowboy_router:constraints()
    , handler     => module()
    , options     => any()
-   , metadata    => map()
+   , metadata    => metadata()
    }.
 -export_type([trail/0]).
 
--type trails() :: [ trails:trail() | cowboy_router:route_path() ].
+%% exported from cowboy_router.erl
```
Comment (elbrujohalcon) [inline src/trails.erl]: _not_ exported ;)

File: src/trails.erl [hunk 2]
```diff
@@ -13,35 +13,52 @@
...
+-type route_match() :: '_' | iodata().
+-type route_path() :: {Path::route_match(), Handler::module(), Opts::any()}
+  | {Path::route_match()
+    , cowboy_router:constraints()
+    , Handler::module()
+    , Opts::any()}.
+-type route_rule() :: {Host::route_match(), Paths::[route_path()]}
+  | {Host::route_match(), cowboy_router:constraints(), Paths::[route_path()]}.
+%%
```
Comment (elbrujohalcon) [inline src/trails.erl]: useless comment line

File: src/trails.erl [hunk 3]
```diff
@@ -52,25 +69,25 @@ to_route_path(Trail) when is_map(Trail) ->
 to_route_path(Trail) when is_tuple(Trail) ->
   Trail.
 
--spec trail(cowboy_router:route_match()
-          , module()) -> cowboy_router:route_path().
+-spec trail(route_match()
```
Comment (elbrujohalcon) [inline src/trails.erl]: There is no need for indentation here

File: src/trails.erl [hunk 4]
```diff
@@ -52,25 +69,25 @@ to_route_path(Trail) when is_map(Trail) ->
...
+          , module()) -> trail().
 trail(PathMatch, ModuleHandler) ->
   trail(PathMatch, ModuleHandler, [], #{}, []).
 
--spec trail(cowboy_router:route_match()
+-spec trail(route_match()
           , module()
-          , any()) -> cowboy_router:route_path().
+          , any()) -> trail().
```
Comment (elbrujohalcon) [inline src/trails.erl]: I think you can un-indent this one now

File: src/trails.erl [hunk 5]
```diff
@@ -52,25 +69,25 @@ to_route_path(Trail) when is_map(Trail) ->
...
 trail(PathMatch, ModuleHandler, Options) ->
   trail(PathMatch, ModuleHandler, Options, #{}, []).
 
--spec trail(cowboy_router:route_match()
+-spec trail(route_match()
           , module()
           , any()
-          , map()) -> cowboy_router:route_path().
+          , map()) -> trail().
```
Comment (elbrujohalcon) [inline src/trails.erl]: I think you can un-indent this one now

Comment (elbrujohalcon) [inline src/trails.erl]: And the last one is not a `map()`, it's `metadata()`

File: src/trails.erl [hunk 6]
```diff
@@ -13,35 +13,52 @@
...
+%%
+
+-type trails() :: [ trails:trail() | route_path() ].
 -export_type([trails/0]).
 
--spec single_host_compile([cowboy_router:route_path()]) ->
+-type method() :: get | put | post | delete | patch | head | options.
+-export_type([method/0]).
+
+-type metadata() :: #{method() => map()}.
```
Comment (elbrujohalcon) [inline src/trails.erl]: this type should be exported.

File: src/trails.erl [hunk 7]
```diff
@@ -52,25 +69,25 @@ to_route_path(Trail) when is_map(Trail) ->
...
-          , map()) -> cowboy_router:route_path().
+          , map()) -> trail().
 trail(PathMatch, ModuleHandler, Options, MetaData) ->
   trail(PathMatch, ModuleHandler, Options, MetaData, []).
 
--spec trail(cowboy_router:route_match()
+-spec trail(route_match()
           , module()
           , any()
           , map()
```
Comment (elbrujohalcon) [inline src/trails.erl]: bad indentation

File: test/trails_SUITE.erl [hunk 8]
```diff
@@ -180,7 +183,31 @@ basic_metadata(_Config) ->
...
-                , {private_file, "index.html"}
+                , {private_file, "index1.html"}
+                , Metadata
+                , []),
+  Metadata = trails:metadata(Trail),
+  {comment, ""}.
+
+ -spec put_metadata(config()) -> {atom(), string()}.
+put_metadata(_Config) ->
+ Metadata = #{ put => #{ description => "Put method"}},
```
Comment (elbrujohalcon) [inline test/trails_SUITE.erl]: indent

File: test/trails_SUITE.erl [hunk 9]
```diff
@@ -180,7 +183,31 @@ basic_metadata(_Config) ->
...
+                , cowboy_static
+                , {private_file, "index2.html"}
+                , Metadata
+                , []),
+  Metadata = trails:metadata(Trail),
+  {comment, ""}.
+
+-spec post_metadata(config()) -> {atom(), string()}.
+post_metadata(_Config) ->
+ Metadata = #{ post => #{ description => "Post method"}},
```
Comment (elbrujohalcon) [inline test/trails_SUITE.erl]: indent

**Pull request merged**